### ISSUE TITLE
Cursor movement between WYSIWYG/expanded modes

### ISSUE THREAD

**Issue by beorn** [typora/typora-issues#2271]
**TLDR; There are some surprises (read usability issues) when moving from WYSIWYG mode into markdown-code-expanded mode, especially when using cursor movements, and these can be alleviated by moving the cursor insertion point to a location that would be more consistent with the WYSIWYG view, not the markdown code view.  Furthermore, it's not handled consistently across links and headlines, for example.** 

I have a suggestion that would at least make editing link text more intuitive for me.  Let’s say you have a link (markdown code expanded) as follows:

	This [a link](http://actual.link.com/) to somewhere.

which will render (in WYSIWYG mode) as:

  **This is _a link_ to somewhere.**

If you click inside of the link title the cursor will be positioned into the link title, as expected (and the markdown code will be expanded, which I don't have a problem with).

If, however, you were moving the cursor to the left from the **to somewhere** into the **_a link_**, the markdown code is shown and the cursor appears to jump to a different location (outside of the link location) then what you saw just a moment ago.  While I can understand this from a “you are editing markdown code” point-of-view, I think it’s a bit suboptimal if you consider the WYSIWYG point-of-view:

1) If you imagine a WYSIWYG editor, you’d expect to be positioned to edit the **_a link_** text.
2) It would better match the pointer-tapping behavior.
3) Multiple delete events (delete or backspace) destroys the markup symbols instead of the text (which gives confusing results).
4) It’s probably much more common to edit link text, so it would make sense to optimize for that.
5) Accidental edits of actual links (which can be quite cryptic) tend to be harder to notice (only way is to click on them, and even then you couldn’t be sure) than accidental edits of link texts (which are obvious).
6) This would make it consistent with how cursoring into some other markdown code sections, such as headlines (```#``` etc) works.  If you move the cursor to the right into a headline it’ll position the cursor to the left of the text, not to the left of (outside of) the markdown code.

I think a potentially better behavior is this:

  - If you’re outside of the link so it’s rendered WYSIWYG (not expanded)
    - If you move to the left from “to somewhere” you would jump directly to the right of “a link” after expanding the markdown code
    - If you move to the right from “This is” you would jump directly to the left of “a link” after expanding the markdown code

  - With the markdown code expanded, however, you can:
    - Use the cursor to move into the link itself and edit it
    - Delete/edit the brackets (since you can see them)

So any accidental deletes/backspaces from outside the link into the link would impact the link text, not the actual link.

The dual WYSIWYG/markdown-code modes of Typora is a cool & unique feature of the editor, but it comes with some compromises.  I think the above changes will improve usability, perhaps alleviating some other complaints too as an alternative to removing this duality and go completely WYSIWYG with popups (refs #1317).

**Comment by beorn** [comment]
Another add-on feature request is to have ``Cmd-K`` select the entire link when you're inside a link.  This makes it still possible to use ``Cmd-K`` to reasonably edit links, not just insert them.

**Comment by abnerlee** [comment]
duplicate to #1313?

**Comment by beorn** [comment]
It's a duplicate in as much as it talks about the link usability issues, but the take on it is different: focus on left/right cursor movements, and there's a suggestion how to at least fix this a bit.
